File: C++/151. Reverse Words in a String.cpp
class Solution {
public:
    string reverseWords(string s) {
        vector<string> a;
        string r;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == ' ') {
                if (!r.empty()) {
                    a.push_back(r);
                    r = "";
                }
            } else {
                r = r + s[i];
            }
        }
        if(!r.empty())
        a.push_back(r);
        string r1;
        for (int i = a.size() - 1; i >= 0; i--) {
            r1 = r1 + a[i];
            if(i!=0)
            r1=r1+" ";
        }
        return r1;
    }
};

File: JAVA/151. Reverse Words in a String.java
class Solution {
    public String reverseWords(String s) {
        int i = 0, j = s.length() - 1;
        String r = "", s1 = "";
        while (i <= j && s.charAt(i) == ' ') {
            i++;
        }
        while (j >= i && s.charAt(j) == ' ') {
            j--;
        }
        while (i <= j) {
            char c = s.charAt(i);
            if (c == ' ') {
                if (!r.isEmpty()) {
                    s1 = r + (s1.isEmpty() ? "" : " ") + s1; 
                    r = ""; 
                }
                while (i <= j && s.charAt(i) == ' ') {
                    i++;
                }
                continue;
            }
            r += c;
            i++;
        }
        if (!r.isEmpty()) {
            s1 = r + (s1.isEmpty() ? "" : " ") + s1;
        }
        return s1;
    }
}
